operating_systems/ls_clone.c

```diff
@@ -0,0 +1,119 @@
+#include <dirent.h> // contains directory-related functions and the `dirent` struct
+#include <sys/stat.h> // contains `stat` struct and `S_ISDIR` macro
+#include <stdbool.h> // contains the definition for booleans
+#include <stdio.h>
+#include <unistd.h> // contains the `optind` variable
+#include <string.h> // for `strlen` and `strcmp`
+
+// Global Default Variables for `ls` flags
+bool exclude_hidden = true;
+bool recursive_ls = false;
+
+int is_directory(const char *path) {
+  struct stat stat_buffer;
+
+  // if the path cannot be read, return False
+  if (stat(path, &stat_buffer) != 0) {
+    return 0;
+  }
+
+  // arrow `->` to access if the variable is a pointer
+  // dot `.` to access if the variable is not a pointer
+  // (*x).foo is equivalent to x->foo
+  return S_ISDIR(stat_buffer.st_mode);
+}
+
+int ls_files(const char *path) {
+  DIR *directory_ptr = opendir(path);
+
+  if (directory_ptr == NULL) {
+    printf("Could not open directory %s\n", path);
+    return 1;
+  }
+
+  while (1) {
+    struct dirent *entry;
+
+    entry = readdir(directory_ptr);
+    if (entry == NULL) {
+      break;
+    }
+
+    // hidden files start with `.`
+    if (exclude_hidden && entry->d_name[0] == '.') {
+      continue;
+    }
+
+    int buffer_size = strlen(path) + strlen(entry->d_name) + 2;
+    char filepath[buffer_size];
+    snprintf(filepath, buffer_size, "%s/%s", path, entry->d_name);
+
+    #define TERMINAL_CYAN  "\x1B[36m"
+    #define TERMINAL_MAGENTA "\x1B[35m"
+    #define TERMINAL_RESET "\x1B[0m"
+
+    if (is_directory(filepath)) {
+      printf("%s%s%s\n", TERMINAL_CYAN, filepath, TERMINAL_RESET);
+    } else {
+      printf("%s\n", filepath);
+    }
+
+    if (recursive_ls == true && is_directory(filepath) && strcmp(entry->d_name, ".") != 0 && strcmp(entry->d_name, "..") != 0) {
+      printf("\n%s%s:\n%s", TERMINAL_MAGENTA, filepath, TERMINAL_RESET);
+      ls_files(filepath);
```

Comment:
Really nice application of algorithms knowledge!

Here's one filesystem quirk to watch out for: what if someone created a symbolic link?

```shell
% ln -s . link-to-self

% ./ls_clone -aR .
Flag `-a` passed. Will list hidden files.
Given option R. Will ls recursively.

./.
./..
./ls_clone
./link-to-self

./link-to-self:
./link-to-self/.
./link-to-self/..
./link-to-self/ls_clone
./link-to-self/link-to-self

./link-to-self/link-to-self:
./link-to-self/link-to-self/.
./link-to-self/link-to-self/..
./link-to-self/link-to-self/ls_clone
./link-to-self/link-to-self/link-to-self

./link-to-self/link-to-self/link-to-self:
./link-to-self/link-to-self/link-to-self/.
./link-to-self/link-to-self/link-to-self/..
./link-to-self/link-to-self/link-to-self/ls_clone
./link-to-self/link-to-self/link-to-self/link-to-self

...
```